plugins/catalog-backend/src/service/AuthorizedLocationService.ts

```diff
@@ -0,0 +1,113 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { LocationSpec, Location, Entity } from '@backstage/catalog-model';
+import { NotAllowedError, NotFoundError } from '@backstage/errors';
+import {
+  catalogLocationCreatePermission,
+  catalogLocationDeletePermission,
+  catalogLocationReadPermission,
+} from '@backstage/plugin-catalog-common';
+import {
+  AuthorizeResult,
+  PermissionAuthorizer,
+} from '@backstage/plugin-permission-common';
+import { LocationService } from './types';
+
+export class AuthorizedLocationService implements LocationService {
+  constructor(
+    private readonly locationService: LocationService,
+    private readonly permissionApi: PermissionAuthorizer,
+  ) {}
+
+  async createLocation(
+    spec: LocationSpec,
+    dryRun: boolean,
+    options?: {
+      authorizationToken?: string;
+    },
+  ): Promise<{
+    location: Location;
+    entities: Entity[];
+    exists?: boolean | undefined;
+  }> {
+    const authorizationResponse = (
+      await this.permissionApi.authorize(
+        [{ permission: catalogLocationCreatePermission }],
+        { token: options?.authorizationToken },
+      )
+    )[0];
+
+    if (authorizationResponse.result === AuthorizeResult.DENY) {
+      throw new NotAllowedError();
+    }
+
+    return this.locationService.createLocation(spec, dryRun);
+  }
+
+  async listLocations(options?: {
+    authorizationToken?: string;
+  }): Promise<Location[]> {
+    const authorizationResponse = (
+      await this.permissionApi.authorize(
+        [{ permission: catalogLocationReadPermission }],
+        { token: options?.authorizationToken },
+      )
+    )[0];
+
+    if (authorizationResponse.result === AuthorizeResult.DENY) {
+      return [];
+    }
+
+    return this.locationService.listLocations();
+  }
+
+  async getLocation(
+    id: string,
+    options?: { authorizationToken?: string },
+  ): Promise<Location> {
+    const authorizationResponse = (
+      await this.permissionApi.authorize(
+        [{ permission: catalogLocationReadPermission }],
+        { token: options?.authorizationToken },
+      )
+    )[0];
+
+    if (authorizationResponse.result === AuthorizeResult.DENY) {
+      throw new NotFoundError(`Found no location with ID ${id}`);
+    }
+
+    return this.locationService.getLocation(id);
+  }
+
+  async deleteLocation(
+    id: string,
+    options?: { authorizationToken?: string },
+  ): Promise<void> {
+    const authorizationResponse = (
+      await this.permissionApi.authorize(
+        [{ permission: catalogLocationDeletePermission }],
+        { token: options?.authorizationToken },
+      )
```

Comment:
I think it's fairly obvious that this isn't powerful enough and we need to pass on more information in the future, right? What does future change look like?

I'm assuming attributes would be a poor way to communicate additional information in this case? With the exception being if we can do something like an `owner` attribute? Another option I assume is to introducing a resource type for locations, would that be a fairly simple change to introduce? Are there any other options I'm missing?

Not looking to stir up too much discussion on what solutions we could go for in the future, mostly interested in how big if a breaking change it would be to introduce each of them

Reply:
I actually don't think it would be a breaking change at all if we add filtering by some attributes - you'd just add whatever attributes and have to additionally handle the `CONDITIONAL` result. Cc @mtlewis

Reply:
> I think it's fairly obvious that this isn't powerful enough and we need to pass on more information in the future, right? What does future change look like?

I'd love to know a bit more about the use cases this doesn't cover. Are there concrete cases where users need to be able to see specific locations, for example? 

For things like deletion of locations, I'm wondering whether it's possible for us to push more of the responsibility for authorization into the catalog backend itself (e.g. the API will delete a location automatically when deleting a corresponding entity, but it's considered a side-effect of the entity deletion, rather than an operation initiated by (and therefore authorized in the context of) the current user.

Reply:
Will merge and we can continue the conversation in the thread offline.